In [2]:
import numpy as np 
import pandas as pd

In [45]:
df = pd.concat(map(pd.read_csv, ['Brazil.csv', 'germany.csv','germany1.csv', 
                                 'Portugal1.csv', 'Spain1.csv', 'Switzerland.csv',
                                'Switzerland1.csv', 'France1.csv',
                                'Brazil1.csv', 'Netherlands.csv', 
                                'Belgium.csv', 'USA.csv', 'England.csv',
                                'Argentina.csv', 'Uruguay.csv', 
                                'Senegal.csv', 'Ecuador.csv', 'CostaRica.csv',
                                'Mexico.csv', 'France.csv', 'Denmark.csv',
                                'Australia.csv', 'Serbia.csv',
                                'Cameroon.csv', 'Ghana.csv', 'Iran.csv',
                                'Poland.csv', 'SaudI.csv', 'Tunisia.csv', 
                                 'Japan.csv', 'Croatia.csv', 'Morocco.csv',
                                'USA1.csv', 'Spain2.csv', 'Portugal2.csv', 
                                 'England2.csv', 'Belgium2.csv', 'Uruguay2.csv',
                                'Canada2.csv', 'Netherlands2.csv', 'Saudi2.csv', 
                                'Croatia2.csv', 'Cameroon2.csv', 'Denmark2.csv',
                                'Costa2.csv', 'Morocco2.csv', 'Poland2.csv',
                                'Tunisia2.csv', 'Japan2.csv', 'Argentina2.csv', 
                                 'Italy.csv', 'South_Korea.csv', 'Nigeria.csv',
                                'Colombia.csv', 'Paraguay.csv', 'Russia.csv',
                                'Algeria.csv', 'Chile.csv', 'Yugoslavia.csv',
                                'Scotland.csv']))



In [46]:
# 158 with wc 2022 teams

In [47]:
df.shape

(214, 17)

In [48]:
df

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  progpasses_z  \
0   0.423077  0.146923  -0.136154     -0.030000  0.429231      0.037692   
1   0.216923  0.348462   0.031538      0.158462  0.835385      0.626923   
2   0.113846  0.392308   0.599231      0.513846  0.833077      0.405385   
3   0.479231  0.609231   0.227692      0.450769  0.770769      0.042308   
4   0.877692  0.773846   0.428462      0.659231  0.754615      0.335385   
..       ...       ...        ...           ...       ...           ...   
2   0.173077  0.019231  -0.180769     -0.154615 -0.334615     -0.305385   
0   0.724615  0.216154   0.162308      0.543077  0.106154      0.170769   
1  -0.235385 -0.176154   0.450769      0.628462  0.112308      0.389231   
2  -0.063077  0.332308   0.383846      0.306154 -0.310769     -0.283846   
3  -0.103077  0.187692   0.225385      0.247692  0.449231      0.306154   

    takeons_z  progruns_z  tackles_z  interceptions_z  clearances_z  blocks_z  \
0    0.244615   -0.220000   0.216154         0.270000     -0.076923 -0.097692   
1    0.270000    0.266923   1.143846         0.834615     -0.059231 -0.134615   
2    0.521538    1.139231   1.109231         0.859231      0.084615  0.103077   
3    0.337692    0.927692   0.506923         1.015385      0.020769  0.381538   
4    0.023077    0.638462   0.493846         0.637692     -0.117692 -0.033846   
..        ...         ...        ...              ...           ...       ...   
2   -0.186154   -0.043846  -0.296923        -0.510769      0.129231 -0.087692   
0   -0.110769   -0.146923  -0.592308         0.101538     -0.200000  0.498462   
1   -0.335385   -0.196923   0.030769        -0.306154     -0.343077 -0.291538   
2   -0.388462   -0.383077  -0.373846        -0.473846     -0.273846 -0.309231   
3   -0.149231   -0.097692   0.577692        -0.180000      0.073077 -0.158462   

    aerials_z   fouls_z  fouled_z    nsxg_z  results  
0   -0.020000 -0.224615  0.100769 -0.124615        2  
1    0.196923 -0.016923  0.033846  0.146923        1  
2    0.683846  0.213846  0.353846  0.786923        9  
3    0.038462  0.039231  0.022308  0.692308        5  
4    0.572308 -0.016154 -0.096923  0.890769        5  
..        ...       ...       ...       ...      ...  
2   -0.153846 -0.136154  0.246154 -0.297692       10  
0    0.317692  0.038462 -0.234615  0.306154       15  
1   -0.146154 -0.166923 -0.230000  0.233846       19  
2    0.197692 -0.378462 -0.386154  0.280000       18  
3    0.555385 -0.006154  0.364615 -0.033846       27  

[214 rows x 17 columns]

In [49]:
df.corr()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  \
goals_z          1.000000  0.595125   0.048376      0.299546  0.144501   
xg_z             0.595125  1.000000   0.373832      0.643746  0.302172   
crosses_z        0.048376  0.373832   1.000000      0.635127  0.230595   
boxtouches_z     0.299546  0.643746   0.635127      1.000000  0.351809   
passes_z         0.144501  0.302172   0.230595      0.351809  1.000000   
progpasses_z     0.011788  0.089852   0.258722      0.221029  0.756853   
takeons_z        0.236356  0.333821   0.115748      0.199458  0.353896   
progruns_z       0.166672  0.390392   0.268164      0.381864  0.578487   
tackles_z        0.199744  0.354681   0.209609      0.380180  0.742328   
interceptions_z  0.170020  0.291017   0.191669      0.282822  0.743682   
clearances_z    -0.187727 -0.392148  -0.239566     -0.435594 -0.118199   
blocks_z         0.022513  0.024446   0.009211      0.113180  0.401582   
aerials_z        0.063352  0.045942   0.242042      0.235632  0.001433   
fouls_z         -0.132051 -0.074164   0.015508     -0.060239  0.039001   
fouled_z         0.041996  0.148063   0.162979      0.149906  0.296412   
nsxg_z           0.348530  0.642626   0.652312      0.848261  0.481358   
results         -0.613551 -0.304032   0.035259     -0.200198 -0.116540   

                 progpasses_z  takeons_z  progruns_z  tackles_z  \
goals_z              0.011788   0.236356    0.166672   0.199744   
xg_z                 0.089852   0.333821    0.390392   0.354681   
crosses_z            0.258722   0.115748    0.268164   0.209609   
boxtouches_z         0.221029   0.199458    0.381864   0.380180   
passes_z             0.756853   0.353896    0.578487   0.742328   
progpasses_z         1.000000   0.145477    0.321025   0.524043   
takeons_z            0.145477   1.000000    0.536393   0.350326   
progruns_z           0.321025   0.536393    1.000000   0.520686   
tackles_z            0.524043   0.350326    0.520686   1.000000   
interceptions_z      0.561778   0.363018    0.421566   0.693822   
clearances_z        -0.026959  -0.111814   -0.055791  -0.212569   
blocks_z             0.364365   0.170438    0.288772   0.249716   
aerials_z            0.041734  -0.106537   -0.035734   0.063998   
fouls_z              0.188255   0.093709    0.024436  -0.007107   
fouled_z             0.188064   0.307633    0.235063   0.297825   
nsxg_z               0.354943   0.391129    0.502926   0.478675   
results             -0.041658  -0.160295   -0.187539  -0.173387   

                 interceptions_z  clearances_z  blocks_z  aerials_z   fouls_z  \
goals_z                 0.170020     -0.187727  0.022513   0.063352 -0.132051   
xg_z                    0.291017     -0.392148  0.024446   0.045942 -0.074164   
crosses_z               0.191669     -0.239566  0.009211   0.242042  0.015508   
boxtouches_z            0.282822     -0.435594  0.113180   0.235632 -0.060239   
passes_z                0.743682     -0.118199  0.401582   0.001433  0.039001   
progpasses_z            0.561778     -0.026959  0.364365   0.041734  0.188255   
takeons_z               0.363018     -0.111814  0.170438  -0.106537  0.093709   
progruns_z              0.421566     -0.055791  0.288772  -0.035734  0.024436   
tackles_z               0.693822     -0.212569  0.249716   0.063998 -0.007107   
interceptions_z         1.000000     -0.120105  0.309376   0.014162  0.111927   
clearances_z           -0.120105      1.000000  0.256962   0.253328  0.235242   
blocks_z                0.309376      0.256962  1.000000   0.054539  0.194611   
aerials_z               0.014162      0.253328  0.054539   1.000000  0.082841   
fouls_z                 0.111927      0.235242  0.194611   0.082841  1.000000   
fouled_z                0.314582      0.066626  0.104951   0.018920  0.245655   
nsxg_z                  0.413978     -0.429796  0.151469   0.130650 -0.001407   
results                -0.110266      0.085575 -0.077828  -0.030329  0.122961   

                 fo

In [50]:
df.shape

(214, 17)

In [51]:
df.duplicated().value_counts()

False    214
dtype: int64

In [52]:
# df.to_csv('WC-stats.csv', index=False)


In [34]:
from sklearn.linear_model import LinearRegression
from sklearn.inspection import permutation_importance
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge



In [35]:
df.head()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  progpasses_z  \
0  0.423077  0.146923  -0.136154     -0.030000  0.429231      0.037692   
1  0.216923  0.348462   0.031538      0.158462  0.835385      0.626923   
2  0.113846  0.392308   0.599231      0.513846  0.833077      0.405385   
3  0.479231  0.609231   0.227692      0.450769  0.770769      0.042308   
4  0.877692  0.773846   0.428462      0.659231  0.754615      0.335385   

   takeons_z  progruns_z  tackles_z  interceptions_z  clearances_z  blocks_z  \
0   0.244615   -0.220000   0.216154         0.270000     -0.076923 -0.097692   
1   0.270000    0.266923   1.143846         0.834615     -0.059231 -0.134615   
2   0.521538    1.139231   1.109231         0.859231      0.084615  0.103077   
3   0.337692    0.927692   0.506923         1.015385      0.020769  0.381538   
4   0.023077    0.638462   0.493846         0.637692     -0.117692 -0.033846   

   aerials_z   fouls_z  fouled_z    nsxg_z  results  
0  -0.020000 -0.224615  0.100769 -0.124615        2  
1   0.196923 -0.016923  0.033846  0.146923        1  
2   0.683846  0.213846  0.353846  0.786923        9  
3   0.038462  0.039231  0.022308  0.692308        5  
4   0.572308 -0.016154 -0.096923  0.890769        5

In [36]:
y = df['results']

In [37]:
X = df.drop(columns=['results'])

In [38]:
X_train, X_test, y_train, y_test = train_test_split(
     X, y, test_size=0.20, random_state=42)

In [39]:
model = Ridge(alpha=1e-2).fit(X_train, y_train)
model.score(X, y)

0.42193954003807654

In [40]:
r = permutation_importance(model, X, y, n_repeats=30, random_state=0)

# Permutation feature importance

In [41]:
scoring = ['r2', 'neg_mean_absolute_percentage_error', 'neg_mean_squared_error']
r_multi = permutation_importance(
     model, X, y, n_repeats=30, random_state=0, scoring=scoring)

for metric in r_multi:
     print(f"{metric}") 
     r = r_multi[metric]
     for i in r.importances_mean.argsort()[::-1]:
            if r.importances_mean[i] - 2 * r.importances_std[i] > 0:
                print(f"    {X.columns[i]:<8}"
                      f"{r.importances_mean[i]:.3f}"
                      f" +/- {r.importances_std[i]:.3f}")
                
                

r2
    goals_z 0.916 +/- 0.089
    xg_z    0.034 +/- 0.013
    boxtouches_z0.032 +/- 0.010
    crosses_z0.025 +/- 0.010
    progruns_z0.023 +/- 0.010
neg_mean_absolute_percentage_error
    goals_z 0.727 +/- 0.109
    fouled_z0.037 +/- 0.014
    boxtouches_z0.035 +/- 0.013
    fouls_z 0.017 +/- 0.006
    tackles_z0.013 +/- 0.006
neg_mean_squared_error
    goals_z 72.476 +/- 7.035
    xg_z    2.656 +/- 0.997
    boxtouches_z2.546 +/- 0.767
    crosses_z1.990 +/- 0.828
    progruns_z1.807 +/- 0.782


- df.shape = (158, 17)
- Ridge Regression = 0.42629810615686115
- r2
    - goals_z 1.282 +/- 0.149
    - progruns_z0.096 +/- 0.029
    - xg_z    0.095 +/- 0.030
    - takeons_z0.026 +/- 0.013
- neg_mean_absolute_percentage_error
    - goals_z 0.988 +/- 0.115
- neg_mean_squared_error
    - goals_z 108.981 +/- 12.672
    - progruns_z8.139 +/- 2.494
    - xg_z    8.047 +/- 2.509
    - takeons_z2.195 +/- 1.067


- df.shape = (125, 17)

- Ridge Regression = 0.4494483686544971

- r2
    - goals_z 0.982 +/- 0.123
    - progruns_z0.132 +/- 0.039
    - xg_z    0.056 +/- 0.017
    - boxtouches_z0.046 +/- 0.023
    - nsxg_z  0.039 +/- 0.013
    - aerials_z0.031 +/- 0.014
- neg_mean_absolute_percentage_error
    - goals_z 0.888 +/- 0.122
    - progruns_z0.181 +/- 0.059
- neg_mean_squared_error
    - goals_z 89.125 +/- 11.206
    - progruns_z11.964 +/- 3.505
    - xg_z    5.055 +/- 1.543
    - boxtouches_z4.180 +/- 2.062
    - nsxg_z  3.509 +/- 1.182
    - aerials_z2.782 +/- 1.244

- df.shape = (145, 17)
- Ridge Regression = 0.43985733944194594

- r2
    - goals_z 0.919 +/- 0.122
    - progruns_z0.077 +/- 0.028
    - nsxg_z  0.069 +/- 0.026
    - boxtouches_z0.034 +/- 0.015
- neg_mean_absolute_percentage_error
    - goals_z 0.817 +/- 0.115
    - progruns_z0.124 +/- 0.036
    - tackles_z0.060 +/- 0.022
    - boxtouches_z0.048 +/- 0.023
- neg_mean_squared_error
    - goals_z 78.467 +/- 10.375
    - progruns_z6.549 +/- 2.364
    - nsxg_z  5.854 +/- 2.186
    - boxtouches_z2.943 +/- 1.292

- NMAPE = the higher the better
- NMSE
- R2 =  is a statistical measure of how close the data are to the fitted regression line . It is the percentage of variation (from 0 to 1) explained by the relationship between two variables.. metric that is used to evaluate the performance of a regression-based machine learning model.